File: backend/src/main/java/etf/bg/diplomski/subject/SubjectService.java
package etf.bg.diplomski.subject;

import etf.bg.diplomski.teacher.TeacherToSubjectCountDTO;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

@Service
public class SubjectService {

  @Autowired private SubjectDTOMapper mapper;

  @Autowired private SubjectRepo subjectRepo;

  public Subject addSubject(SubjectDTO subject) {
    if(subjectRepo.findByName(subject.name()).isPresent()){
      throw new SubjectAlreadyExistsException();
    }
    Subject s = new Subject(subject.name(), subject.status(), subject.description());
    return subjectRepo.save(s);
  }

  public List<SubjectDTO> getAllSubjects() {
    List<SubjectDTO> returnList=new ArrayList<>();
    for (Subject subject :subjectRepo.findAll().stream().toList()) {
        SubjectDTO temp = mapper.apply(subject);
        returnList.add(temp);
    }
    return returnList;
  }

  public void deleteSubject(String name) {
    Optional<Subject> temp=subjectRepo.findByName(name);
    Subject subject;
    if(temp.isPresent()) {
      subject = temp.get();
    }else{
      throw new SubjectNotFoundException();
    }
    subjectRepo.delete(subject);
  }

  public Subject getSubjectByName(String name) {
    if (subjectRepo.findByName(name).isPresent()) {
      return subjectRepo.findByName(name).get();
    }
    throw new SubjectNotFoundException();
  }

  public void acceptSubject(String name) throws SubjectNotFoundException{
    Optional<Subject> temp=subjectRepo.findByName(name);
    Subject subject;
    if(temp.isPresent()) {
      subject =temp.get();
      subject.setStatus("APPROVED");
    }else{
      throw new SubjectNotFoundException();
    }
    subjectRepo.save(subject);
  }

  public List<Subject> getAllSubjectRequests() {
    return subjectRepo.findAllSubjectRequests();
  }

  public List<TeacherToSubjectCountDTO> getSubjectTeacherCounts() {
    List<Object[]> results = subjectRepo.findSubjectTeacherCounts();
    return results.stream()
            .map(result -> new TeacherToSubjectCountDTO((String) result[0], (Long) result[1]))
            .toList();
  }

}

File: vue_front/src/services/SubjectService.js
import axios from "axios";

const BACKEND_URL='http://localhost:8084'

class SubjectService{

    searchSubjects(searchParam) {
        const data={
          searchParam:searchParam
        }
        return axios.post(BACKEND_URL+'/subject/v1/searchSubjects',data);
      }
    
      getAllSubjects() {
        return axios.get(BACKEND_URL+'/subject/v1/getAllSubjects');
      }
    
      getAllSubjectRequests() {
        return axios.get(BACKEND_URL+'/subject/v1/getAllSubjectRequests');
      }
    
    
      approveSubject(name) {
        const data={
          name:name
        }
        return axios.post(BACKEND_URL+'/subject/v1/approveSubject',data);
      }
    
      addSubject(name,status,description) {
        const data={
          name:name,
          status:status,
          description:description
        }
        return axios.post(BACKEND_URL+'/subject/v1/addSubject',data);
      }
    
      addSubjectRequest(name,description) {
        const data={
          name:name,
          description:description
        }
        return axios.post(BACKEND_URL+'/subject/v1/addSubjectRequest',data);
      }
    
    
      deleteSubject(name) {
        const data={
          name:name
        }
        return axios.post(BACKEND_URL+'/subject/v1/deleteSubject',data);
      
      }

      GetTeachersPerSubject(){
        return axios.get(BACKEND_URL+'/subject/teacherPerSubjectCounts')
      }
    
      GetTeachersPerGradeLevel(){
        return axios.get(BACKEND_URL+'/subject/getTeachersPerGradeLevel')
      }
}

export default new SubjectService()